apiserver/facades/controller/actionpruner/pruner.go

```diff
@@ -4,37 +4,57 @@
 package actionpruner
 
 import (
+	"github.com/juju/errors"
+
 	"github.com/juju/juju/apiserver/common"
 	apiservererrors "github.com/juju/juju/apiserver/errors"
 	"github.com/juju/juju/apiserver/facade"
 	"github.com/juju/juju/apiserver/params"
 	"github.com/juju/juju/state"
 )
 
+// API provides access to the action pruner API.
 type API struct {
 	*common.ModelWatcher
+	cancel     <-chan struct{}
 	st         *state.State
 	model      *state.Model
 	authorizer facade.Authorizer
 }
 
-func NewAPI(st *state.State, r facade.Resources, auth facade.Authorizer) (*API, error) {
-	m, err := st.Model()
+// NewAPI returns an action pruner API.
+func NewAPI(ctx facade.Context) (*API, error) {
+	m, err := Model(ctx)
 	if err != nil {
-		return nil, err
+		return nil, errors.Trace(err)
 	}
-
 	return &API{
-		ModelWatcher: common.NewModelWatcher(m, r, auth),
-		st:           st,
-		authorizer:   auth,
+		ModelWatcher: common.NewModelWatcher(m, ctx.Resources(), ctx.Auth()),
+		st:           ctx.State(),
+		authorizer:   ctx.Auth(),
+		cancel:       ctx.Cancel(),
 	}, nil
 }
 
+// Model returns the model for a context (override for tests).
+var Model = func(ctx facade.Context) (state.ModelAccessor, error) {
+	m, err := ctx.State().Model()
+	if err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+// Prune performs the action pruner operation (override for tests).
+var Prune = state.PruneOperations
```

Comment:
It's because the testing facade context uses state.State and other such structs directly and there's currently no easy way to get around it. So yeah, it was a pragmatic decision to fit in with what we currently have without changing the world. The testing facade context is used in a lot of places. Fixing it will also allow us to remove the use of the dreaded JujuConnSuite but that will be a lot of work best done in a separate series of PRs.